fastcrypto/src/groups/multiplier/integer_utils.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Given a binary representation of a number in little-endian format, return the digits of its base
+/// `2^bits_per_digit` expansion.
+pub fn compute_base_2w_expansion<const N: usize>(
+    bytes: &[u8; N],
+    bits_per_digit: usize,
+) -> Vec<usize> {
+    assert!(0 < bits_per_digit && bits_per_digit <= usize::BITS as usize);
+
+    // The base 2^window_size expansions digits in little-endian representation.
+    let mut digits = Vec::new();
+
+    // Compute the number of digits needed to represent the numbed in base 2^w. This is equal to
+    // ceil(8*N / window_size), and we compute like this because div_ceil is unstable as of rustc 1.69.0.
+    let digits_count = div_ceil(8 * N, bits_per_digit);
+
+    for i in 0..digits_count {
+        digits.push(get_bits_from_bytes(
+            bytes,
+            bits_per_digit * i,
+            bits_per_digit * (i + 1),
+        ));
+    }
+    digits
+}
+
+/// Get the integer represented by a given range of bits of a byte from start to end (exclusive).
+#[inline]
+fn get_lendian_from_substring(byte: &u8, start: usize, end: usize) -> u8 {
+    assert!(start < end);
+    byte >> start & ((1 << (end - start)) - 1) as u8
+}
+
+/// Compute ceil(numerator / denominator).
+pub(crate) fn div_ceil(numerator: usize, denominator: usize) -> usize {
+    (numerator + denominator - 1) / denominator
+}
+
+/// Get the integer represented by a given range of bits of a an integer represented by a little-endian
+/// byte array from start to end (exclusive). The `end` argument may be arbitrarily large, but if it
+/// is larger than 8*N, the remaining bits of the byte array will be assumed to be zero.
+#[inline]
+pub fn get_bits_from_bytes<const N: usize>(bytes: &[u8; N], start: usize, end: usize) -> usize {
+    assert!(start < end && start < 8 * N);
```

Comment:
Is this something that in Sui can be called arbitrarily by a malicious sender? or we should return an error instead of panicking?

Reply:
All values are determined by constants (window_width and scalar_size) which is set in the secp256r1 file, so a caller has no influence on them.